spacy/pipeline/lemmatizer.py

```diff
@@ -0,0 +1,277 @@
+    def mode(self):
+        return self._mode
+
+    def __call__(self, doc: Doc) -> Doc:
+        """Apply the lemmatizer to one document.
+
+        doc (Doc): The Doc to process.
+        RETURNS (Doc): The processed Doc.
+
+        DOCS: https://spacy.io/api/lemmatizer#call
+        """
+        for token in doc:
+            if self.overwrite or token.lemma == 0:
+                token.lemma_ = self.lemmatize(token)[0]
+        return doc
+
+    def lookup_lemmatize(self, token: Token) -> List[str]:
+        """Lemmatize using a lookup-based approach.
+
+        token (Token): The token to lemmatize.
+        RETURNS (list): The available lemmas for the string.
+        """
+        lookup_table = self.lookups.get_table("lemma_lookup", {})
+        result = lookup_table.get(token.text, token.text)
+        if isinstance(result, str):
+            result = [result]
+        return result
+
+    def rule_lemmatize(self, token: Token) -> List[str]:
+        """Lemmatize using a rule-based approach.
+
+        token (Token): The token to lemmatize.
+        RETURNS (list): The available lemmas for the string.
+        """
+        cache_key = (token.orth, token.pos, token.morph)
+        if cache_key in self.cache:
+            return self.cache[cache_key]
+        string = token.text
+        univ_pos = token.pos_
+        morphology = token.morph.to_dict()
+        if isinstance(univ_pos, int):
+            univ_pos = UPOS_NAMES.get(univ_pos, "X")
+        univ_pos = univ_pos.lower()
+        if univ_pos in ("", "eol", "space"):
+            return [string.lower()]
+        # See Issue #435 for example of where this logic is requied.
+        if self.is_base_form(univ_pos, morphology):
+            return [string.lower()]
+        index_table = self.lookups.get_table("lemma_index", {})
+        exc_table = self.lookups.get_table("lemma_exc", {})
+        rules_table = self.lookups.get_table("lemma_rules", {})
+        if not any(
+            (
+                index_table.get(univ_pos),
+                exc_table.get(univ_pos),
+                rules_table.get(univ_pos),
+            )
+        ):
+            if univ_pos == "propn":
+                return [string]
+            else:
+                return [string.lower()]
+
+        index = index_table.get(univ_pos, {})
+        exceptions = exc_table.get(univ_pos, {})
+        rules = rules_table.get(univ_pos, {})
+        orig = string
+        string = string.lower()
+        forms = []
+        oov_forms = []
+        for old, new in rules:
+            if string.endswith(old):
+                form = string[: len(string) - len(old)] + new
+                if not form:
+                    pass
+                elif form in index or not form.isalpha():
+                    forms.append(form)
+                else:
+                    oov_forms.append(form)
+        # Remove duplicates but preserve the ordering of applied "rules"
+        forms = list(dict.fromkeys(forms))
```

Comment:
Wow, nice trick

Reply:
I didn't touch this part!

Reply:
Huh, so you didn't: https://github.com/explosion/spaCy/commit/955b95cb8ba5185525e736089f7c81743f32efa0

At least I was right that I didn't remember writing it...